### PR TITLE
Ruby on Rails capstone project - Budget app

### PR DESCRIPTION
**Pull request by Aduda-Boaz**
- This is a Ruby on Rails capstone project about building a mobile web application where one can manage his/her budget: one has a list of transactions associated with a category to see what he/she has spent.

- Users are able to:
    - Register and log in, so that the data is private to them.
    - Introduce new transactions associated with a category.
    - See the money spent on each category.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,24 +1,56 @@
...
+## Video presentation
 
-* Configuration
+ > [Click here](https://www.loom.com/share/a1dd7a2d035d40e0b7d29bda9fc9d4b0?t=4)
 
-* Database creation
+## Live Demo
 
-* Database initialization
+ > [Click here](https://rails-capstone-budget-app.herokuapp.com/)
```
Comment (MahmoudBakr23) [inline README.md]: Required

- Please fix the live demo link

File: app/models/category.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class Category < ApplicationRecord
+  belongs_to :user, foreign_key: 'user_id'
+  has_many :category_transactions, dependent: :destroy
```
Comment (MahmoudBakr23) [inline app/models/category.rb]: Required

- As you used the many through association, it's required to have two association types, one for the main model which is the transaction, and one for the third model that connects the category and transaction through it. Currently, you're missing the has many transactions through category transaction association

File: app/models/category_transaction.rb [hunk 3]
```diff
@@ -0,0 +1,4 @@
+class CategoryTransaction < ApplicationRecord
+  belongs_to :category
+  has_many :transaction
```
Comment (MahmoudBakr23) [inline app/models/category_transaction.rb]: Required

-  Following the association issue above, in this third model that connects the two main models to each other should only belong to them. So, please modify the has many association to belongs to

File: app/views/transactions/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+    <div class="form_inputs">
+      <div>
+        <%= f.text_field :name, placeholder: "Name", class: "input"%>
+      </div>
+
+      <div>
+        <%= f.number_field :amount, placeholder: "Price", class: "input"%>
+      </div>
+    </div>
```
Comment (MahmoudBakr23) [inline app/views/transactions/new.html.erb:19]: Required

- In the new transaction form view, please add a field with all the groups names to select from as assignment to this new transaction. This is to make use of the many to many relationship. Currently, the user can only add a transaction to one group, but the project requirements dictate that q user should be able to assign the same transaction to as many groups as wanted, therefore the ERD association will be applied. See the image from the project requirements down below.

Screenshot_2022-08-13-18-59-19-68

**Review by MahmoudBakr23 — CHANGES_REQUESTED** [review]
Hello @Aduda-Boaz 😃, 

Great work and effort you put into this project so far.

## Highlights 🚀 

There are some excellent points that you have made perfectly. Keep up! 💪
- Well done working with controllers and RESTful methods
- Well done working with rails routes
- Well done creating the correct migrations
- Well done adding associations so far 
- Well done setting up the devise gem for authentication
- Well done styling the app according to the given design 
- Well done testing the integrations of your app
- Descriptive pr summary and headline 
- Descriptive Readme and presentation
- Correct Linters 

However, there is always room for improvement.

> Keep in mind that your long hours of work are appreciated. However, the code review process is for improving the code you wrote, and not for criticizing it. 😄

 ## Required Changes ♻️
_Please check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

n/a

Cheers and Happy coding!👏👨‍💻💻

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/category.rb (hunk 2 above)
Comment (Aduda-Boaz) [reply app/models/category.rb]: Added the association type

File: README.md (hunk 1 above)
Comment (Aduda-Boaz) [reply README.md]: I am working on the heroku issue and I feel it will take me some time to fix it.

File: app/models/category_transaction.rb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+class CategoryTransaction < ApplicationRecord
+  belongs_to :category
+end
```
Comment (Deepakdanger) [inline app/models/category_transaction.rb]: - [ ] There should be an association between `transactions` and `category_transaction`. Here it should be `belongs_to :transaction`.

File: app/views/shared/_category_card.html.erb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<div class="category">
+  <div class="category_card">
+    <div class="icon_card">
+      <%= image_tag(category.icon, class: "icon") if category.icon.attached? %>
+    </div>
+    <div class= "category_info">
+      <%= category.name %>
+    </div>
+  </div>
+</div>
```
Comment (Deepakdanger) [inline app/views/shared/_category_card.html.erb]: - [ ] Currently, I'm not able to see the amount. Please show the amount as well of the transaction and in the category index show the total amount of transaction as per the design. (Both are missing the amount)

### Category index
Screenshot from 2022-08-14 00-40-28

### Category show page
Screenshot from 2022-08-14 00-30-03

**Review by Deepakdanger — CHANGES_REQUESTED** [review]
Hi @Aduda-Boaz, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/shared/_category_card.html.erb [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+  <div class="category_card">
+    <div class="icon_card">
+      <%= image_tag(category.icon, class: "icon") if category.icon.attached? %>
+    </div>
+    <div class= "category_info">
+      <%= category.name %>
+      <div class="category_transaction">
+        $<%= category.sum_total %>
+      </div>
+    </div>
```
Comment (uwadonat) [inline app/views/shared/_category_card.html.erb:11]: - At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category). Now the sum of your transactions are not calculated. Kindly work on it.

Your design:
Screenshot from 2022-08-13 23-47-11

Given design:
Screenshot from 2022-08-13 23-48-41

**Review by uwadonat — CHANGES_REQUESTED** [review]
## Hi @Aduda-Boaz, :hand: 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

### To highlight! :+1: 
- Descriptive pull request message ✔️
- Clean and well organized project ✔️
- No linter errors ✔️
almoo

 ## Status: Required Changes ♻️
_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100%.
Please, remember to tag me @uwadonat in your question so I can receive the notification.

------

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

**Review by danielufeli — APPROVED** [review]
Hi @Aduda-Boaz ,

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

a

## Highlights
- Nice commits messages✔️
- Descriptive PR ✔️
- Good use of GitHub flow ✔️
- Clean Code ✔️

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @danielufeli in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**